test/foundry/Tiers.t.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: CAL
+pragma solidity =0.8.17;
+
+import {VaultConfig} from "../../contracts/vault/receipt/ReceiptVault.sol";
+import {CreateOffchainAssetReceiptVaultFactory} from "../../contracts/test/CreateOffchainAssetReceiptVaultFactory.sol";
+import {Test, Vm} from "forge-std/Test.sol";
+import {
+    OffchainAssetReceiptVault,
+    UnauthorizedSenderTier
+} from "../../contracts/vault/offchainAsset/OffchainAssetReceiptVault.sol";
+import {OffchainAssetReceiptVaultFactory} from
+    "../../contracts/vault/offchainAsset/OffchainAssetReceiptVaultFactory.sol";
+import {ReadWriteTier} from "../../contracts/test/ReadWriteTier.sol";
+import {OffchainAssetVaultCreator} from "./OffchainAssetVaultCreator.sol";
+import {ITierV2} from "@rainprotocol/rain-protocol/contracts/tier/ITierV2.sol";
+
+contract TiersTest is Test, CreateOffchainAssetReceiptVaultFactory {
+    event SetERC20Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+    event SetERC1155Tier(address sender, address tier, uint256 minimumTier, uint256[] context, bytes data);
+
+    /// Test setERC20Tier event
+    function testSetERC20Tier(
+        uint256 fuzzedKeyAlice,
+        string memory assetName,
+        string memory assetSymbol,
+        bytes memory fuzzedData,
+        uint8 fuzzedMinTier,
+        uint256[] memory fuzzedContext,
+        address tier
+    ) external {
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+
+        fuzzedMinTier = uint8(bound(fuzzedMinTier, uint256(1), uint256(8)));
+
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        // Create the vault
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+
+        // Grant the necessary role
+        vault.grantRole(vault.ERC20TIERER(), alice);
+
+        // Emit the expected event
+        vm.expectEmit(true, true, true, true);
+        emit SetERC20Tier(alice, address(tier), fuzzedMinTier, fuzzedContext, fuzzedData);
+
+        // Call the function that should emit the event
+        vault.setERC20Tier(address(tier), fuzzedMinTier, fuzzedContext, fuzzedData);
```

Comment:
so yes, we expect to see the event emitted, but where/how do we test that the tier contract was updated in the contract?

where do we test the failure cases where the setERC20Tier does not complete successfully for whatever reason?